tsdb/db_append_v2_test.go

```diff
@@ -0,0 +1,7745 @@
+	require.NoError(t, db.Compact(ctx))
+	require.Equal(t, 1.0, prom_testutil.ToFloat64(db.head.metrics.checkpointCreationTotal))
+
+	// As the data spans for 59 blocks, 58 go to disk and 1 remains in Head.
+	require.Len(t, db.Blocks(), 58)
+	// Though WAL was truncated only once, head should be truncated after each compaction.
+	require.Equal(t, 58.0, prom_testutil.ToFloat64(db.head.metrics.headTruncateTotal))
+
+	// The compaction should have only truncated first 2/3 of WAL (while also rotating the files).
+	first, last, err = wlog.Segments(db.head.wal.Dir())
+	require.NoError(t, err)
+	require.Equal(t, 40, first)
+	require.Equal(t, 61, last)
+
+	// The first checkpoint would be for first 2/3rd of WAL, hence till 39.
+	// That should be the last checkpoint.
+	_, cno, err := wlog.LastCheckpoint(db.head.wal.Dir())
+	require.NoError(t, err)
+	require.Equal(t, 39, cno)
+
+	// Case 2: Old blocks on disk.
+	// The above blocks will act as old blocks.
+
+	// Creating a block to cover the data in the Head so that
+	// Head will skip the data during replay and start fresh.
+	blocks := db.Blocks()
+	newBlockMint := blocks[len(blocks)-1].Meta().MaxTime
+	newBlockMaxt := db.Head().MaxTime() + 1
+	require.NoError(t, db.Close())
+
+	createBlock(t, db.Dir(), genSeries(1, 1, newBlockMint, newBlockMaxt))
+
+	db = newTestDB(t, withDir(db.Dir()), withOpts(opts))
+	db.DisableCompactions()
+
+	// 1 block more.
+	require.Len(t, db.Blocks(), 59)
+	// No series in Head because of this new block.
+	require.Equal(t, 0, int(db.head.NumSeries()))
+
+	// Adding sample way into the future.
+	app = db.AppenderV2(context.Background())
+	_, err = app.Append(0, lbls, 0, blockRange*120, rand.Float64(), nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.NoError(t, app.Commit())
+
+	// The mint of head is the last block maxt, that means the gap between mint and maxt
+	// of Head is too large. This will trigger many compactions.
+	require.Equal(t, newBlockMaxt, db.head.MinTime())
+
+	// Another WAL file was rotated.
+	first, last, err = wlog.Segments(db.head.wal.Dir())
+	require.NoError(t, err)
+	require.Equal(t, 40, first)
+	require.Equal(t, 62, last)
+
+	require.Equal(t, 0.0, prom_testutil.ToFloat64(db.head.metrics.checkpointCreationTotal))
+	require.NoError(t, db.Compact(ctx))
+	require.Equal(t, 1.0, prom_testutil.ToFloat64(db.head.metrics.checkpointCreationTotal))
+
+	// No new blocks should be created as there was not data in between the new samples and the blocks.
+	require.Len(t, db.Blocks(), 59)
+
+	// The compaction should have only truncated first 2/3 of WAL (while also rotating the files).
+	first, last, err = wlog.Segments(db.head.wal.Dir())
+	require.NoError(t, err)
+	require.Equal(t, 55, first)
+	require.Equal(t, 63, last)
+
+	// The first checkpoint would be for first 2/3rd of WAL, hence till 54.
+	// That should be the last checkpoint.
+	_, cno, err = wlog.LastCheckpoint(db.head.wal.Dir())
+	require.NoError(t, err)
+	require.Equal(t, 54, cno)
+}
+
```

Comment:
Maybe add a TODO here to also handle 
```
TestQuerier_ShouldNotPanicIfHeadChunkIsTruncatedWhileReadingQueriedChunks
TestChunkQuerier_ShouldNotPanicIfHeadChunkIsTruncatedWhileReadingQueriedChunks
```

Once we get rid of v1 Appender .  Just to avoid automatically deleting those.